bootstrap/kubeadm/internal/ignition/clc/clc.go

```diff
@@ -0,0 +1,377 @@
+      contents:
+        inline: |
+          ---
+          {{ .KubeadmConfig | Indent 10 }}
+    {{- if .NTP }}{{- if and .NTP.Enabled .NTP.Servers }}
+    - path: /etc/ntp.conf
+      mode: 0644
+      contents:
+        inline: |
+          # Common pool
+          {{- range  .NTP.Servers }}
+          server {{ . }}
+          {{- end }}
+
+          # Warning: Using default NTP settings will leave your NTP
+          # server accessible to all hosts on the Internet.
+
+          # If you want to deny all machines (including your own)
+          # from accessing the NTP server, uncomment:
+          #restrict default ignore
+
+          # Default configuration:
+          # - Allow only time queries, at a limited rate, sending KoD when in excess.
+          # - Allow all local queries (IPv4, IPv6)
+          restrict default nomodify nopeer noquery notrap limited kod
+          restrict 127.0.0.1
+          restrict [::1]
+    {{- end }}{{- end }}
+`
+)
+
+type render struct {
+	*cloudinit.BaseUserData
+
+	KubeadmConfig            string
+	UsersWithPasswordAuth    string
+	FilesystemDevicesByLabel map[string]string
+}
+
+func defaultTemplateFuncMap() template.FuncMap {
+	return template.FuncMap{
+		"Indent":         templateYAMLIndent,
+		"Split":          strings.Split,
+		"Join":           strings.Join,
+		"MountpointName": mountpointName,
+	}
+}
+
+func mountpointName(name string) string {
+	return strings.TrimPrefix(strings.ReplaceAll(name, "/", "-"), "-")
+}
+
+func templateYAMLIndent(i int, input string) string {
+	split := strings.Split(input, "\n")
+	ident := "\n" + strings.Repeat(" ", i)
+	return strings.Join(split, ident)
+}
+
+func renderCLC(input *cloudinit.BaseUserData, kubeadmConfig string) ([]byte, error) {
+	t := template.Must(template.New("template").Funcs(defaultTemplateFuncMap()).Parse(clcTemplate))
+
+	usersWithPasswordAuth := []string{}
+	for _, user := range input.Users {
+		if user.LockPassword != nil && !*user.LockPassword {
+			usersWithPasswordAuth = append(usersWithPasswordAuth, user.Name)
+		}
+	}
+
+	filesystemDevicesByLabel := map[string]string{}
+	if input.DiskSetup != nil {
+		for _, filesystem := range input.DiskSetup.Filesystems {
+			filesystemDevicesByLabel[filesystem.Label] = filesystem.Device
+		}
+	}
+
+	data := render{
+		BaseUserData:             input,
+		KubeadmConfig:            kubeadmConfig,
+		UsersWithPasswordAuth:    strings.Join(usersWithPasswordAuth, ","),
+		FilesystemDevicesByLabel: filesystemDevicesByLabel,
+	}
+
+	var out bytes.Buffer
+	if err := t.Execute(&out, data); err != nil {
+		return nil, errors.Wrapf(err, "failed to render template")
+	}
+
+	return out.Bytes(), nil
+}
+
+// Render renders the provided user data and CLC snippets into Ignition config.
+func Render(input *cloudinit.BaseUserData, clc *bootstrapv1.ContainerLinuxConfig, kubeadmConfig string) ([]byte, string, error) {
+	if input == nil {
+		return nil, "", errors.New("empty base user data")
+	}
+
+	if clc == nil {
```

Comment:
why can't we this nil? isn't it an optional user input?

Reply:
Good point. I think it can be made optional.